lib/storage/drbd.py

```diff
@@ -397,6 +404,22 @@ def _AssembleLocal(self, minor, backend, meta, size):
         base.ThrowError("drbd%d: can't attach local disk: %s",
                         minor, result.output)
 
+    def _WaitForMinorSyncParams():
+      """Call _SetMinorSyncParams and raise RetryAgain on errors.
+      """
+      if self._SetMinorSyncParams(minor, self.params):
+        raise utils.RetryAgain()
+      else:
+        return []
+
+    if self._NeedsLocalSyncerParams():
+      # Retry because disk config for DRBD resource may be still uninitialized.
+      try:
+        utils.Retry(_WaitForMinorSyncParams, 1.0, 5.0)
+      except utils.RetryTimeout:
+        base.ThrowError("drbd%d: can't set the synchronization parameters: %s" %
+                        (minor, utils.CommaJoin(sync_errors)))
```

Comment:
`sync_errors` here is unbound.  Not sure how to propagate the error from `_WaitForMinorSyncParams` when `RetryTimeout` is reached.  Ideas?

Reply:
I am not exactly sure I understood how utils.retry works but the following should preserve the sync_errors variable through the exception:
```python
  for cmd in cmds:
    result = utils.RunCmd(cmd)
    if result.failed:
      base.ThrowError("drbd%d: can't attach local disk: %s",
                      minor, result.output)

  def _WaitForMinorSyncParams():
    """Call _SetMinorSyncParams and raise RetryAgain on errors.
    """
    sync_errors = self._SetMinorSyncParams(minor, self.params)
    if sync_errors:
      raise utils.RetryAgain(sync_errors)
    else:
      return []
      
  if self._NeedsLocalSyncerParams():
    # Retry because disk config for DRBD resource may be still uninitialized.
    try:
      utils.Retry(_WaitForMinorSyncParams, 1.0, 5.0)
    except utils.RetryTimeout as e:
      base.ThrowError("drbd%d: timed out setting the synchronization parameters: %s" %
                      (minor, utils.CommaJoin(e.args[0])))
```

Reply:
Thanks for the hint, I've incorporated this suggestion now.  I tried testing it by setting a very low timeout, but could not trigger the error.  Probably my toy cluster is not busy enough!

I also removed the explicit return from `_WaitForMinorSyncParams` as nothing is checking it.